src/ast/code_generator.ml

```diff
@@ -21,25 +21,67 @@ type type_env =
 
 let build_type_env (m : modul)
   (func_ty : binary param_type * binary result_type)
-  (owi_funcs : (string * int) list) : type_env =
+  (owi_funcs : (string * int) array) : type_env =
   let param_types = List.map snd (fst func_ty) in
-  let global_types (* would get simplified with #353 *) =
+  let global_types =
     List.map
-      (fun (x : (global, binary global_type) Runtime.t Indexed.t) ->
-        match Indexed.get x with
+      (fun (x : (global, binary global_type) Runtime.t) ->
+        match x with
```

Comment:
`function` should now be enough instead of `fun c -> match x with ...`

Reply:
In the definition of binary module fields, record field names are duplicated like:

```ocaml
type exports =
  { global : export list
  ; mem : export list
  ; table : export list
  ; func : export list
  }

type global =
  { typ : binary global_type (* TODO: init : binary+const expr*)
  ; init : binary expr
  ; id : string option
  }

...

type data =
  { id : string option
  ; init : string
  ; mode : data_mode
  }

...

type elem =
  { id : string option
  ; typ : binary ref_type (* TODO: init : binary+const expr*)
  ; init : binary expr list
  ; mode : elem_mode
  }
```

So without the type annotation OCaml cannot infer the type

Reply:
No, I think in `List.map f l` OCaml determines the (erroneous) type of `f` first then uses that to check `l`. So it thinks `m.global` should have type of `(elem, (_ * binary heap_type)) Runtime.t array`

Reply:
OK fair enough.